### ISSUE TITLE
Use MSG_DONTWAIT (instead of O_NONBLOCK) for tty fds?

### ISSUE THREAD

**Issue by bnoordhuis** [libuv/libuv#3602]
Refs: https://github.com/nodejs/node/issues/42826

Libuv dups fds 0-2 when they refer to a tty because setting O_NONBLOCK affects other processes it shares those file descriptors with.

The logic is pretty hairy (is this a pty slave? is this... etc.) Maybe we can sidestep that completely by switching to sendmsg() and recvmsg() with the MSG_DONTWAIT flag.

**Comment by vtjnash** [comment]
Duplicate of https://github.com/libuv/libuv/issues/2062

**Comment by bnoordhuis** [comment]
Maybe? That issue seems to be about a number of things but I'll add a `Fixes #2062`.

**Comment by vtjnash** [comment]
It covers a related set of problems with nodejs from passing fd 0-2 to libuv. I typically suggest applications do not that, and instead pass dup(0), etc, as it avoids this problem and other like it.

MSG_DONTWAIT seems very interesting, but will it work for non-socket types? I can't find any statements online (though it should be fairly easy to change libuv and run CI to check)

**Comment by bnoordhuis** [comment]
> MSG_DONTWAIT seems very interesting, but will it work for non-socket types?

The answer turned out to be 'no'. Closing as a wontfix then.

**Comment by Keno** [comment]
As of Linux 6.4, it appears that IOCB_NOWAIT (reachable as RWF_NOWAIT from pwritev2 as far as I can tell) works for pipes. The O_NONBLOCK issue is still causing problems for non-tty fds. I think it would be worthwhile investigating if we can avoid setting the flag and use `RWF_NOWAIT` instead.

https://github.com/torvalds/linux/commit/c04fe8e32f907ea668f3f802387c1148fdb0e6c9

**Comment by bnoordhuis** [comment]
Libuv ostensibly still supports linux 3.10; 6.4 is - almost to the day! - 10 years newer, 2013 vs. 2023.

I have no problem with bumping the minimum supported version but then we'd probably pick 5.10 or 5.14; 6.4 is too new.

We can feature-detect support at runtime but that makes the logic inside libuv more complex and it doesn't help people stuck on older kernels. I'll have to think about whether that's worthwhile.

**Comment by Keno** [comment]
I really would like to get to a point where libuv stops setting O_NONBLOCK on fds passed to it from the outside or that it gives to other processes (well it already does the latter, but that breaks libuv's assumptions). We (julia) have about half a dozen issues open about this causing unexpected behavior (we got another two just this week, which caused me to look at this again). It causes issues in both directions. If the fd is shared with a process that is not prepared to handle EAGAIN, it can cause errors there. And likewise if it is shared with a process that likes to reset O_NONBLOCK (which bash does), it'll break libuv's assumptions about the behavior of the system call. In my ideal world, we would stop changing this flag and 1) trying something like RWF_NOWAIT or if not 2) extend the re-opening logic to other kinds of fds (the example we got this week was a fifo, which could be re-opened also) and then 3) if all else fails kicking the io off to the thread pool.

There's also the issue that (apparently) O_NONBLOCK isn't actually guaranteed to not block. According to some LKML sleuthing it'll block in some corner cases (but RWF_WAIT will not), in particular when operating on a pipe that's being splice'd (that's what that commit series above was about) into or on metadata-update of the mtime of a fifo.

**Comment by bnoordhuis** [comment]
> that it gives to other processes

You mean file descriptors inherited by child processes? Or SCM_RIGHTS? When/where does libuv set/neglect to clear O_NONBLOCK?

**Comment by Keno** [comment]
> You mean file descriptors inherited by child processes?

Yes

> When/where does libuv set/neglect to clear O_NONBLOCK?

libuv clears it, but this can cause problems when the same fd is still used by the parent. 

https://github.com/JuliaLang/julia/pull/60326 https://github.com/JuliaLang/julia/issues/60141 https://github.com/JuliaLang/julia/issues/24440 are the related julia issues, but I'd need to turn it into a libuv-only reproducer to be more precise.

**Comment by bnoordhuis** [comment]
> libuv clears it, but this can cause problems when the same fd is still used by the parent

Ah, I get it now. Okay, that's a good point in favor of RWF_NOWAIT. You'd still have issues on other platforms though, right?

**Comment by Keno** [comment]
Yeah, it's not a perfect solution, but if we do the threadpool fallback, we can maybe at least make it correct everywhere and then argue for per-syscall nonblock versions on other OSes also to restore the performance.

**Comment by bnoordhuis** [comment]
I tested it today but RWF_NOWAIT only works with pipes; not with fifos<sup>1</sup>, or ttys<sup>1</sup>, or regular files<sup>2</sup>. That's not very useful.

If the way libuv currently does things isn't satisfactory, then it's probably best to perform reads and writes from a separate thread.

<sup>1</sup> fails with EOPNOTSUPP
<sup>2</sup> fails with EAGAIN

**Comment by Keno** [comment]
> not with fifos

My understanding is that this is due to needing to update the timestamp. I think the series at https://lore.kernel.org/all/20251120064859.2911749-1-hch@lst.de/ would allow that to be enabled.

> ttys

I think this could be fixed with appropriate refactoring in the kernel. My impression is that the RWF_NOWAIT work is primarily driven by io_uring requirements, for which ttys are probably not a priority would be my guess.

> regular files

Probably the same timestamp issue.

> If the way libuv currently does things isn't satisfactory, then it's probably best to perform reads and writes from a separate thread.

I think this fallback is required in any case, either for older kernel versions or for corner cases like a fifo being on a file system that doesn't support lazy timestamp updates.

**Comment by Keno** [comment]
I think another plausible answer is that we should not add the RWF_NOWAIT special case and instead focus on https://github.com/libuv/libuv/issues/4044 plus adding the threadpool fallback.

**Comment by Keno** [comment]
We suggested to the user to try extending the re-opening hack we have for ttys to pipes (which works on linux, but not other unices I've checked). Unfortunately, their pipes were created by a privileged process, so the re-opening fails. I've asked @xal-0 to take a look at this. I think we need the threadpool fallback in any case. I had hoped that for linux we could maybe just always re-open, but given the permissions issue that might not be feasible. Pending a full migration to io_uring, perhaps there's a place for RWF_NOWAIT after all.
